### ISSUE TITLE
bookmarks moved

### ISSUE THREAD

**Issue by lazymonkey2** [floccusaddon/floccus#1767]
### Which version of floccus are you using?

5.3.3

### How many bookmarks do you have, roughly?

1400

### Are you using other means to sync bookmarks in parallel to floccus?

no

### Sync method

Google Drive

### Which browser are you using? In case you are using the phone App, specify the Android or iOS version and device please.

firefox, edge, chrome

### Describe the Bug

I sync bookmarks between my three browsers (merge).
after a sync I found that some bookmarks are moved from one folder to a different one:
- in firefox I have 13 bookmarks in a certain folder.
- in edge the bookmarks are synced correctly
- in chrome the 13 bookmarks are moved in a different folder
- the name of the original folder in firefox and the destination (wrong) folder in chrome is the same -> ai  (is it relevant?)
- even after multiple syncs the bookmarks in chrome are not moved to the correct folder
- to find differences I exported bookmarks from the three browsers in html format, then I removed some useless data in the html files (like timestamps and icons data) and then I compared the html files using winmerge.

### Expected Behavior

synced bookmarks correctly

### To Reproduce

I don't know.  It's just happening when I sync my bookmarks.
I'll attach a screenshot of the difference in html files, and redacted logs.

![bug bookmarks moved](https://github.com/user-attachments/assets/deb63d93-b923-4d57-ac30-945f8399899c)

[floccus-5.3.3-2024-11-13-redacted-firefox.log](https://github.com/user-attachments/files/17734073/floccus-5.3.3-2024-11-13-redacted-firefox.log)

[floccus-5.3.3-2024-11-13-redacted-edge.log](https://github.com/user-attachments/files/17734076/floccus-5.3.3-2024-11-13-redacted-edge.log)

[floccus-5.3.3-2024-11-13-redacted-chrome.log](https://github.com/user-attachments/files/17734077/floccus-5.3.3-2024-11-13-redacted-chrome.log)

### Debug log provided

- [X] I have provided a debug log file

**Comment by marcelklehr** [comment]
> When two folders with the same name are moved to a different location in the same sync run, there is currently no way for the diff algorithm to differentiate them and they may get mixed up

I've since implemented a similarity measure that helps the diff algorithm detect which folder is more likely to match when multiple folders have the same name and there is no cache, so this issue should not happen anymore. Thank you again for reporting! :blue_heart:

**Comment by lazymonkey2** [comment]
hello, 
I've got this issue again, on floccus v5.4.0, so please take a look at it.
it happened again on two folders with the same name, but nested into two different subfolders.

I think the diff engine should use the full path of folders to be able to differentiate two folders with same name.
in this last case if the diff had used the full path it probably have synced correctly.

this should work as expected except maybe when there are two folders with the same name and located in the same subfolder.

**Comment by marcelklehr** [comment]
Damn

> I think the diff engine should use the full path of folders to be able to differentiate two folders with same name.

Mmh, I don't think this is going to work as the issue is about diffing moved folders, so I believe this happens when folders are moved. Can you confirm this?

I will look into this again.

**Comment by lazymonkey2** [comment]
> I believe this happens when folders are moved. Can you confirm this?

no. in my specific case the folders with same name are not moved.

what is going wrong for me is when I add a new bookmark to an existing folder, which by chance has the same name as another folder, but these two folders are located into different subfolders, like this example:

1st folder full path:
/toolbar/dev/security

2nd folder full path:
/toolbar/pc/security

for my specific issue these two "security" folders are already present in both pc 1 and pc 2.
but in a more general case it's possibile that the folder on pc 2 is not present, and must be created by sync engine.

so on pc 1 I add a new bookmark to first "security" folder, then on pc 2 after sync I find the new bookmark added to the 2nd folder instead of the first one.

I don't know the internals of your sync engine, but I still believe that the full path can univocally identify a folder or a bookmark. 

with the exception of two different folders with same path:

/toolbar/pc/security
/toolbar/pc/security

(security here is a folder).

it's possible to create two folders with same name in the same subfolder (just tested in chrome and firefox).
maybe you can tell them apart by looking at the order in which they appear (like "first security folder" and "second security folder").

Another problematic case could be two bookmarks with the same name and url in the same folder.

/toolbar/pc/security/new_bookmark
/toolbar/pc/security/new_bookmark

(here new_bookmark is a bookmark, not a folder).

it's possible to create (probably unintentionally) two identical bookmarks in the same folder by using drag and drop.

Maybe the bookmark APIs have access to other meta information, like for example the creation timestamp? could it be useful?
but that infomation could be local to each different pc, so I'm not sure it could be used to identifiy different objects with same name or url.

**Comment by marcelklehr** [comment]
> so on pc 1 I add a new bookmark to first "security" folder, then on pc 2 after sync I find the new bookmark added to the 2nd folder instead of the first one.

Can you try resetting the cache in floccus via "Trigger sync from scratch"? It could be that there's still a wrong mapping going on between the two folders that was never removed even though the establishing of the mapping is now prevented.

**Comment by marcelklehr** [comment]
Floccus uses IDs to uniquely identify bookmarks and folders. To track which local folder is equivalent to which server folder there are mapping records. These could be set in the wrong way in previous versions, though. This is why I'm asking you to reset the cache and sync from scratch to reestablish those mapping records.

**Comment by cjpalermo21** [comment]
Yes  this happens all the time ... I realized it and began renaming my folders something unique (because, if there are two folders with the same name, *all* the bookmarks get dumped into one of them).

**Comment by marcelklehr** [comment]
> all the bookmarks get dumped into one of them

You mean all bookmarks from all folders? :thinking:

**Comment by marcelklehr** [comment]
Yes, I've tried to fix this in v5.4.0. The effects of that fix might
only be visible after resetting the cache once, though.

**Comment by marcelklehr** [comment]
Closing  this as I consider it fixed. Please comment here or open a new issue if you continue to experience this in 2025.

**Comment by cjpalermo21** [comment]
Thanks Marcel - it seems that deleting the file from the google drive and
resetting the cache - when there's an issue - seems to clear it up (so
far)! :P
